source/Dig/base.util.data.bmx

```diff
@@ -220,232 +220,523 @@ Type TData
+		Return data.Contains(TLowerString.Create(key))
+	End Method
 
-		'only react if the "::" is in the middle of something
-		If groupsEnabled
-			Local pos:Int = ls.Find("::")
-			If pos > 0
-				Local group:String = Left(ls.orig, pos)
-				Local groupData:TData = TData(Get(group))
-				If groupData
-					Return groupData.Get(Right(ls.orig, pos+1), defaultValue)
-				EndIf
-			EndIf
+
+	Method Has:Int(key:TLowerString)
+		If Not self.data Then Return False
+		If Not key Then Return False
+
+		Return data.Contains(key)
+	End Method
+
+
+	Method Get:Object(key:String, defaultValue:Object=Null)
+		If Not self.data Then Return Null
+		If Not key Then Return defaultValue
+
+		Local ls:TLowerString = TLowerString.Create(key)
+		Local result:Object = self.data.ValueForKey(ls)
+		If result Then Return Result
+
+		If defaultValue And Not self.data.Contains(ls)
+			Return defaultValue
 		EndIf
+		
+		Return Null
+	End Method
+
+
+	Method Get:Object(key:TLowerString, defaultValue:Object=Null)
+		If Not self.data Then Return Null
+		If Not key Then Return defaultValue
+
+		Local result:Object = self.data.ValueForKey(key)
+		If result Then Return Result
 
-		Local result:Object = data.ValueForKey(ls)
-		If result Then
-			Return result
+		If defaultValue And Not self.data.Contains(key)
+			Return defaultValue
+		EndIf
+		
+		Return Null
+	End Method
+
+
+	Method GetString:String(key:String, defaultValue:String=Null)
+		If Not self.data Then Return Null
+		If Not key Then Return defaultValue
+
+		Local ls:TLowerString = TLowerString.Create(key)
+		Local result:Object = self.data.ValueForKey(ls)
+		If result Then Return result.ToString()
+		
+		'check if the entry is just "empty" (skip lookup is defaultValue is also Null)
+		If defaultValue And Not self.data.Contains(ls) 
+			Return defaultValue
+		EndIf
+		
+		Return ""
+	End Method
+
+
+	Method GetString:String(key:TLowerString, defaultValue:String=Null)
+		If Not self.data Then Return Null
+		If Not key Then Return defaultValue
+
+		Local result:Object = self.data.ValueForKey(key)
+		If result Then Return result.ToString()
+		
+		'check if the entry is just "empty" (skip lookup is defaultValue is also Null)
+		If defaultValue And Not self.data.Contains(key) 
+			Return defaultValue
+		EndIf
+		
+		Return ""
+	End Method
+
+
+	Method GetBool:Int(key:String, defaultValue:Int=False)
+		If Not self.data Then Return Null
+		If Not key Then Return defaultValue
+
+		Local ls:TLowerString = TLowerString.Create(key)
+		Local result:Object = self.data.ValueForKey(ls)
+		'empty = default
+		If Not result Then Return defaultValue
+		
+		Local s:String = String(result)
+		'string interpretation
+		If s
+			s = s.ToLower()
+			Select s
+				Case "true", "yes"
+					Return True
+				Default
+					'also allow "1" / "1.00000" or "33"
+					If Int(s) >= 1 Then Return True
+					Return False
+			End Select
+		'number interpretation
+		ElseIf TDoubleData(result) and Int(TDoubleData(result).value) >= 1
```

Comment:
wasn't the interpretation that everything other than 0.0 is true?

Reply:
also ist auch 0.00001 True? Ich habe das als Integer interpretiert ... man kann nun argumentieren, dass hier gerundet werden sollte. Aber wie gesagt, ist 0.00001 schon True, oder erst "ab 0.5" ?

Reply:
Ich wollte keine Änderung sondern nur fragen, ob sich dadurch was ändert...

Reply:
Naja der alte Code war:
```
	Method GetBool:Int(key:Object, defaultValue:Int=False)
		Local result:Object = Get(key)
		If Not result Then Return defaultValue
		Select String(result).toLower()
			Case "true", "yes"
				Return True
			Default
				'also allow "1" / "1.00000" or "33"
				If Int(String(result)) >= 1 Then Return True
				Return False
		End Select
		Return False
```
Es wurde also frueher sogar immer als False ausgewertet, egal welcher Wert drin war!

Beispiel:
```
SuperStrict
Framework Brl.StandardIO

Type TDoubleData 'von TData extrahiert)
	Field value:Double

	Method ToString:String()
		Return String(value)
	End Method
End Type


Local d:TDoubleData = New TDoubleData
d.value = 1.2345

Local o:Object = d

Print "String(o) = " + String(o)
Print "o.ToString() = " + o.ToString()
```

Ausgabe:
```
String(o) = 
o.ToString() = 1.2344999999999999
```
String(x) ist hier nur ein Cast zu einem String - und da es ein Objekt ist und kein String, bleibt das Ergebnis leer.
Nur mit "ToString()" wuerde per Default (wenn nicht selbst implementiert) die Speicheradresse zurueckgegeben werden - oder wie hier, der eigentliche Wert.

Zusammengefasst:
Alte Version konnte nur Strings interpretieren, Rest war automatisch "False" (und das war nicht so gewollt)

Heisst aber, das "alte Verhalten" (falls GetBool dort auf Double/Int so angewendet) weicht vom jetzigen ab. Allerdings wurde original immer `AddBoolString` genutzt - und das wertete den Wert als String aus und verglich mit `"1"`.
Das bedeutet, dass nun wiederum die "Auswertungsergebnisse" gleich sein sollten.